webapp/src/workspace.ts

```diff
@@ -721,6 +721,29 @@ export async function commitAsync(hd: Header, options: CommitOptions = {}) {
     }
 }
 
+export async function restoreCommitAsync(hd: Header, commit: pxt.github.CommitInfo) {
+    await cloudsync.ensureGitHubTokenAsync();
+
+    const files = await getTextAsync(hd.id)
+    const gitjsontext = files[GIT_JSON]
+    const gitjson = JSON.parse(gitjsontext) as GitJson
+    const parsed = pxt.github.parseRepoId(gitjson.repo)
+    const date = new Date(Date.parse(commit.committer.date));
+    const restored: pxt.github.CreateCommitReq = {
+        message: lf("Restore '{0} {1}'", date.toLocaleString(), commit.message),
+        parents: [gitjson.commit.sha],
+        tree: commit.tree.sha
+    }
+
+    const commitId = await pxt.github.createObjectAsync(parsed.fullName, "commit", restored)
+    await pxt.github.fastForwardAsync(parsed.fullName, parsed.tag, commitId, true)
```

Comment:
I think there should be no force here.

Reply:
It does fail saying an update is not a fast forward

Reply:
If it fails, it means the parent of the commit you're creating is not the current head.

Reply:
In other words, if it fails and your force it, you *will* lose commits